### ISSUE TITLE
Shelly2 ghost switching with shutter switch

### ISSUE THREAD

**Issue by chrizza87** [arendst/Sonoff-Tasmota#4724]
**Describe the bug**
_There is a ghost switching for shelly2 device. I installed the shelly2 behind a jarolift THOR shutter switch (link: https://www.jalousiescout.de/Rolladen-Zeitschaltuhr/Jarolift-Thor-Rollladen-Zeitschaltuhr.html). If i use the shelly2 profile the shelly2 relay switches at a certain time continuously (like 5 times per second). I also tried the generic device with Switch1 and Switch2, also with Switch1n and Switch2n. With the n it was even worse because, there the shelly switched once without a switch attached. 
I tried the original shelly2 firmware at first. There was no ghost switching, but the attached jarolift switch didn't work properly. The switch state was not recognized corretly. Thats why a flashed the tasmota firmware. There the switch works perfectly, but i have this annoying ghost switching, which only stops when i disconnect the switch. Sometimes it works perfectly for a day and the out of the blue the relay switches continuously
I activated the debug mode for logging and there i saw that the switch is the src of the switching on/off._

_Also, make sure these boxes are checked [x] before submitting your issue - Thank you!_
- [x] _Searched the problem in issues and in the wiki_
- [x] _Hardware used_ : Shelly2
- [x] _Development/Compiler/Upload tools used_ : Arduino Board and Platform IO for Windows
- [x] _You have tried latest release or development binaries?_ : yes
- [x] _Provide the output of command_``status 0`` :
```
14:17:23 MQT: livingroom/windowShutters/stat/STATUS = {"Status":{"Module":18,"FriendlyName":["livingRoomWindowShutters","livingRoomWindowShutters2"],"Topic":"windowShutters","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS1 = {"StatusPRM":{"Baudrate":115200,"GroupTopic":"sonoffs","OtaUrl":"xxxxxxxx","RestartReason":"Software/System restart","Uptime":"1T15:08:58","StartupUTC":"2018-12-22T22:08:25","Sleep":50,"BootCount":18,"SaveCount":328,"SaveAddress":"F4000"}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS2 = {"StatusFWR":{"Version":"6.4.0.1(sonoff)","BuildDateTime":"2018.12.17 21:45:20","Boot":31,"Core":"2_4_2","SDK":"2.2.1(cfd48f3)"}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["XXXXXXX",""],"TelePeriod":300,"SetOption":["0000C009","558180C0","00000000"]}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS4 = {"StatusMEM":{"ProgramSize":522,"Free":480,"Heap":21,"ProgramFlashSize":1024,"FlashSize":2048,"FlashChipId":"1540EF","FlashMode":3,"Features":["00000407","0FDAE794","240383A0","23B617CE","00003BC0"]}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS5 = {"StatusNET":{"Hostname":"windowShutters-3511","IPAddress":"192.168.178.67","Gateway":"192.168.178.1","Subnetmask":"255.255.255.0","DNSServer":"192.168.178.1","Mac":"CC:50:E3:13:4D:B7","Webserver":2,"WifiConfig":4}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS6 = {"StatusMQT":{"MqttHost":"xxx.xxx.xxx.xxx","MqttPort":1883,"MqttClientMask":"livingRoomWindowShutters_%06X","MqttClient":"livingRoomWindowShutters_134DB7","MqttUser":"DVES_USER","MqttType":1,"MAX_PACKET_SIZE":1000,"KEEPALIVE":15}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS7 = {"StatusTIM":{"UTC":"Mon Dec 24 13:17:23 2018","Local":"Mon Dec 24 14:17:23 2018","StartDST":"Sun Mar 25 02:00:00 2018","EndDST":"Sun Oct 28 03:00:00 2018","Timezone":"+01:00","Sunrise":"08:42","Sunset":"16:57"}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS10 = {"StatusSNS":{"Time":"2018-12-24T14:17:23","Switch1":"ON","Switch2":"ON"}}
14:17:23 MQT: livingroom/windowShutters/stat/STATUS11 = {"StatusSTS":{"Time":"2018-12-24T14:17:23","Uptime":"1T15:08:58","Vcc":3.586,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"FRITZ!Box 7390 CS","BSSId":"24:65:11:50:BE:20","Channel":4,"RSSI":76}}}
```

**To Reproduce**
_Steps to reproduce the behavior: Flash shelly2 with tasmota, attach the mentioned switch (can happen with another one too. This id don't know) and wait...._

**Expected behavior**
_Shelly2 works like the sonoff devices without ghost switching the relay continuously after a certain time_

**Additional context**
_Switch is attached like described in the shelly docu. 
N -> N
L -> L 
O1 -> Shutter_Down
O2 -> Shutter_Up
S1 -> Switch_down
S2 -> Switch_up
._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by arendst** [comment]
Thx for the info. Needs investigation but I'm afraid I need info from Allterco as I think additional GPIO's may be involved but I do not own the schematics.

**Comment by chrizza87** [comment]
Okay thanks for the quick reply. I hope you can gain some knowledge 👍 I will wait an update ;-)

**Comment by arendst** [comment]
BTW you might want to experiment with the switch debounce delay. it is currently set to 50 milliseconds. 

Use command ``switchdebounce 100`` to change it to a less sensitive value which might work better in your case. The value can change up to 1000 milliseconds.

As you also noticed the n option makes it worse, due to not using input pullup, I suggest to go back to module Shelly 2 (which uses pullup) and try the above setting.

Pls let us know if it makes any difference.

**Comment by chrizza87** [comment]
okay thank you. I will try it and report back

**Comment by chrizza87** [comment]
Hey @arendst, 
after one day so far it works. There was no more ghost switching until now. I will investigate it.

I added a section for this topic in the shelly wiki: https://github.com/arendst/Sonoff-Tasmota/wiki/Shelly-1-and-2

Hope this is okay ;-) I will close this issue, but it would be great if you can keep me posted if you have new scematics from Allterco.

Thanks in advance and happy holidays!

**Comment by chrizza87** [comment]
Now after 26 hours the issue has come back -.-

**Comment by arendst** [comment]
But does it come back once or more often?

If so, increase the debounce and test again.

**Comment by chrizza87** [comment]
sorry was on tour on the holidays. It switched several times Up and Down. It stopped when a detached the switch again. I will try your proposal.

**Comment by ascillato2** [comment]
@chrizza87 please, try with latest Tasmota from today from http://thehackbox.org/tasmota/

**Comment by chrizza87** [comment]
thx @ascillato2, i will try it

**Comment by ascillato** [comment]
@chrizza87 please, try with latest Tasmota from today (6.4.1.2) (new updates from Theo) from http://thehackbox.org/tasmota/

**Comment by chrizza87** [comment]
thx @ascillato2 i updated the firmware and will report back!

**Comment by andreasloe** [comment]
I might have the same problem using a device where GPIO 0 to 3 are used for relays and buttons (for a circuit diagram see [https://forum.fhem.de/index.php?action=dlattach;topic=89818.0;attach=104885;image](this link). When I press switch2 it says
`16:45:16 MQT: stat/eltako/RESULT = {"POWER2":"ON"}`
`16:45:16 MQT: stat/eltako/POWER2 = ON`
`16:45:16 APP: Knopf2 Mehrfachdruck 1` <== This is strange, I pressed only once!
`16:45:16 CFG: in Flash gespeichert am F9, zählen 147, Bytes 3584`
`16:45:16 SRC: Button`
`16:45:16 MQT: stat/eltako/RESULT = {"POWER2":"OFF"}`
`16:45:16 MQT: stat/eltako/POWER2 = OFF`
`16:45:17 WIF: Prüfe Verbindung...`
`16:45:17 WIF: verbunden`
`16:45:17 CFG: in Flash gespeichert am F8, zählen 148, Bytes 3584`
`16:45:37 WIF: Prüfe Verbindung`...
`16:45:37 WIF: verbunden`
I tried several debounce times to no avail. Checking the new firmware now!

**Comment by andreasloe** [comment]
PS Here is a [https://www.bilder-upload.eu/bild-4827c3-1546106398.png.html](Screenshot of the configuration)

**Comment by andreasloe** [comment]
It did not work with the newest release
`19:17:03 HTP: Main Menu`
`19:17:05 SRC: WebGui from 192.168.2.10`
`19:17:05 MQT: stat/eltako/RESULT = {"POWER2":"ON"}`
`19:17:05 MQT: stat/eltako/POWER2 = ON`
`19:17:05 APP: Button2 multi-press 1`  <== nope, pressed only once
`19:17:06 CFG: Saved to flash at F8, Count 182, Bytes 3584`
`19:17:06 SRC: Button`
`19:17:06 MQT: stat/eltako/RESULT = {"POWER2":"OFF"}`
`19:17:06 MQT: stat/eltako/POWER2 = OFF`
`19:17:07 CFG: Saved to flash at F7, Count 183, Bytes 3584`
`19:17:11 HTP: Console`
`19:17:11 WIF: Checking connection...`
`19:17:11 WIF: Connected`
`19:17:14 HTP: Main Menu`

**Comment by chrizza87** [comment]
ghost switching is still happening. not mutiple switches after another. Now just switching once.

I have upgraded to the latest firmware and set the debounce rate back to 50

**Comment by arendst** [comment]
If still ghost switching try the following:

1) Use the latest dev release (v6.4.1.3) as of 20181230
2) Select Shelly 2 as module
3) Set SwitchDebounce to 51. This will enable a feature that will keep current state 1 as long as there are no state change 0 probes within 50 mSec.
4) If no solution try Set SwitchDebounce to 52. This will enable a feature that will keep current state 0 as long as there are no state change 1 probes within 50 mSec.
5) If still fails set SwitchDebounce to 100. This will disable the previous options from bullet 3 and 4 but increase probe samples to 10.

This may take some time for testing but try to report back what worked best.

BTW I cannot test this as my Shelly 2 works as expected without any ghost switching while using very short leads to the switches and no "strange" loads.

**Comment by chrizza87** [comment]
Hey @arendst, i tried the third approach. so far there was ghost switching just once and i discovered, that this was after a restart of the module. It had set the PowerOnState to 0 ({"PowerOnState":0}), but the relay 2 was set to ON after the restart.

**Comment by arendst** [comment]
Great.

Could you find in the log if relay 2 was set to ON externally (by retained message)?

**Comment by chrizza87** [comment]
i must add an logserver an my raspberry. i just have the console in the webinterface right now.

In the last 30 minutes the module has restarted 2 times, first due to "Hardware Watchdog" and  second due to "Software Watchdog". 

The second relay was switched on in both restarts. I tried cut the power off and force a restart, there the relay hasn't switched.

Very strange.

**Comment by ascillato** [comment]
Hi, the log can be seen in the console of Tasmota in the webui. No need to set up a syslog for that test. 

You can use the command `weblog 4` to have more debug information. If you hace a retained message you will see SRC:MQTT so the source of your command is mqtt. You will see SRC:SWITCH if the command came from the switch.

The 2 restarts you suffer is bad flashing. You shouldn't have any type of watchdog restarts. Sorry but you need to erase all flash with esptool.py (as explained in the wiki) and flash again with esptool.py the precompiled bins from development branch at http://thehackbox.org/tasmota/

**Comment by chrizza87** [comment]
but you can just see the current logs, when you open the page. I took me one minute after the ghost switching to look at http page of the module. Or is there a way too see older logs?

Okay i updated the firmware 2 days ago via "firmware update". I will erase the flash and flash it again.

**Comment by ascillato** [comment]
The retained messages happens only when the device boots up, so after the device is on, you go to the console and you can see there the SRC: as previous post.

**Comment by chrizza87** [comment]
okay thank you. i will try this too. :-)

**Comment by chrizza87** [comment]
this is what a get after restart

``
CFG: Loaded from flash at F5, Count 370
00:00:00 SRC: Restart
00:00:00 Project sonoff livingRoomDoorShutters Version 6.4.1.4(1fb7993-sonoff)-2_4_2
00:00:00 SRC: Switch
00:00:00 RSL: livingroom/doorShutters/stat/RESULT = {"POWER1":"ON"}
00:00:00 RSL: livingroom/doorShutters/stat/POWER1 = ON
00:00:00 RSL: livingroom/doorShutters/stat/RESULT = {"POWER1":"OFF"}
00:00:00 RSL: livingroom/doorShutters/stat/POWER1 = OFF
00:00:00 SRC: Switch
00:00:00 RSL: livingroom/doorShutters/stat/RESULT = {"POWER2":"ON"}
00:00:00 RSL: livingroom/doorShutters/stat/POWER2 = ON
00:00:00 CFG: Saved to flash at F4, Count 371, Bytes 3584
``

**Comment by ascillato** [comment]
Mmmm, may be the issue is just noise coming from your switch. Can you try a normal wall switch?

**Comment by arendst** [comment]
The switch seems to have all kind of feedback possibilities like "motoranfrier". This may be the cause of the switch commands you see as tries to actuate the motor.

You might want to try to change the switch for a "normal" pushbutton Up/Down and see if these "ghosts" switches still happen.

**Comment by naldoxable** [comment]
I think it is an issue of the last release, I'm experiencing the same on all my Sonoffs, (Basic and Dual). Randomly turning on for few ms and then off, multiple times during the day

**Comment by ascillato2** [comment]
Ok. I will put the issue ON HOLD up to you have time to test another pushbutton. Thanks

**Comment by MarcG046** [comment]
I am also experiencing ghost switching issues and I wonder if they could be related to software rather than hardware.

I have connected a Tuya curtain switch to a WEMOS with Sonoff-Tasmota (6.5.0 release version)
I removed the Tuya's internal controller and wifi board and took wires from the 3.3v power supply, the relay drivers and the touch keys outside to an external WEMOS.

Unfortunately the output impedance of the touch key chip seems rather high, because it seems unable to sufficiently pull the internal pull-ups of the WEMOS low.
I have turned off the pull-ups by configuring the switch inputs as 'switchn' versions. Then everything works.
Except that there are sporadic ghost switches. Roughly once a week the device switches by itself (there is a 'ESP-SRC: Switch' in the logs)

Due to the high impedance on the wire it may be susceptible to interference. 
Still, I wonder if the cause may be in software instead of hardware.

I have set the SwitchDebounce to 403. If I understand correctly that means that an interfering pulse would have to last for over 400ms to cause a switch. It seems unlikely that this would happen (?)
It also seems that for lower debounce values ghost switching occur equally frequently (although I do not have statistically significant data)
